aip/0194.md

```diff
@@ -45,18 +46,21 @@ The following codes **should not** be retried for any request:
 
 - `OK`: The request succeeded.
 - `CANCELLED`: An application can cancel a request, which **must** be honored.
-- `DEADLINE_EXCEEDED`: An application can set a deadline, which **must** be honored.
-- `INVALID_ARGUMENT`: Retrying a request with an invalid argument will never succeed.
-- `DATA_LOSS`: This is an unrecoverable error and **must** immediately be surfaced to the application.
+- `DEADLINE_EXCEEDED`: An application can set a deadline, which **must** be
+  honored.
+- `INVALID_ARGUMENT`: Retrying a request with an invalid argument will never
+  succeed.
+- `DATA_LOSS`: This is an unrecoverable error and **must** immediately be
+  surfaced to the application.
 
 ### Generally non-retryable codes:
 
 The following codes generally **should not** be retried for any request:
 
 - `RESOURCE_EXHAUSTED`: This code may be a signal that quota is exhausted.
-  Retries therefore may not be expected to work for several hours; meanwhile the
-  retries may have billing implications. If `RESOURCE_EXHAUSTED` is used for
-  other reasons than quota and the expected time for the resource to become
+  Retries therefore may not be expected to work for several hours; meanwhile
+  the retries may have billing implications. If `RESOURCE_EXHAUSTED` is used
+  for other reasons than quota and the expected time for the resource to become
   available is much shorter, it may be retryable.
 - `INTERNAL`: This code generally means that some internal part of the system
```

Comment:
For complicated systems serving at high QPS, I believe that a certain (low) percentage of requests failing for "unexplained" reasons is relatively expected and relatively acceptable. I think it's actually quite reasonable to retry all of INTERNAL, UNKNOWN, and ABORTED at a higher level (and possibly at the RPC level).

In the case of our system, a customer-meaningful chunk of work requires 10s of thousands of RPCs or more, so in the absence of a reasonable retry policy it's virtually guaranteed that users will see failures.

Reply:
The problem is that the client does not have enough information to know that the retry is safe. In many cases, they may be sending a duplicate request that (for example) creates a second resource.

It would be reasonable to configure a gRPC service config for this if you know something about a particular endpoint, but it is not something we should do for every RPC.

Reply:
Perhaps it would be useful to provide slightly more nuanced guidance here, something like "Generally not retriable, but some APIs may choose to document this as retriable if they meet requirements XYZ"?

Reply:
Well, the problem is that "documenting as retriable" does nothing -- automated processes have to handle it. So really the API should just publish a gRPC service config that makes that RPC be retried, and then downstream tooling does not do anything.